xla/service/gpu/fusions/transforms/simplify_arith.cc

```diff
@@ -228,6 +229,24 @@ struct RewriteTruncExtShuffle : public OpRewritePattern<mlir::gpu::ShuffleOp> {
   }
 };
 
+static std::optional<Interval> GetSelectRange(mlir::Operation* sel) {
+  // Match |x| implemented as (x >= 0) ? x : (0 - x).
+  auto x = mlir::matchers::m_Val(sel->getOperand(1));
+  if (!sel->getOperand(1).getType().isInteger(32) ||
+      !mlir::matchPattern(
+          sel, mlir::m_Op<mlir::arith::SelectOp>(
+                   mlir::m_Op<mlir::arith::CmpIOp>(x, mlir::m_Zero()), x,
+                   mlir::m_Op<mlir::arith::SubIOp>(mlir::m_Zero(), x)))) {
+    return std::nullopt;
+  }
+  if (sel->getOperand(0).getDefiningOp<mlir::arith::CmpIOp>().getPredicate() !=
+      CmpIPredicate::sge) {
+    return std::nullopt;
+  }
+  // Annotate |x| as >= 0.
```

Comment:
We should probably refine the range we return here with the range of `x`. Also, in general, let's get the range of both operands and return the union.

Reply:
Included the range of x. Didn't understand about "both operands": x is the only input to the expression matched here.

Reply:
Sorry, I meant when the pattern doesn't match:

select(pred, x, y) -> union(getrange(x), getrange(y))

in the pattern, we should get the abs of range of x and then intersect with what you compute. I think something like this should be right:

```
Interval result ...
if (x_range.has_value()) {
  auto positive_range = x_range.max(*x_range, {0, 0});
  auto abs_negative_range = -x_range.min(*y_range, {0, 0});
  auto abs_x_range = positive_range.Union(abs_negative_range);
  return result.Intersect(abs_x_range);
} 
return result;
```

Reply:
I see. I think my change is still valid (it annotates only the exact |x| pattern) and you can add your more general logic for select on top.

Reply:
Taking the union of x's range and the nonnegative integers is not correct, no. But feel free to just remove that part again.

Reply:
Indeed. range(|x|) = {0, x.upper}, right? Then should it just be
```
  if (x_range.has_value()) {
    return result.Intersect(*x_range);
  }
```
?

Reply:
Indeed!

Reply:
I added a couple more tests with x's range.

Reply:
Thanks!